### ISSUE TITLE
Build failure with util-linux master

### ISSUE THREAD

**Issue by asavah** [systemd/systemd#8507]
### Submission type

 - Bug report

Note: not sure whether this is systemd ot util-linux bug, or both, 
please close if the bug does not belong to systemd.

### systemd version the issue has been seen with

master @ https://github.com/systemd/systemd/commit/0d9fca76bb69e162265b2d25cb79f1890c0da31b

### Used distribution

custom, arm, cross compiled

### In case of bug report: Expected behaviour you didn't see

Build succeeds

### In case of bug report: Unexpected behaviour you saw

Build fails
```[111/1410] Compiling C object 'src/core/core@sta/service.c.o'.
FAILED: src/core/core@sta/service.c.o
armv8-rpi3-linux-gnueabihf-gcc -pipe -Wno-unused-value -Wno-unused-but-set-variable -Wno-deprecated-declarations -Wno-misleading-indentation -Wno-unused-function -Wno-unused-result -Wno-unused-parameter -Wimplicit-fallthrough=0 -Wno-unused-but-set-parameter -Wno-int-in-bool-context -mcpu=cortex-a53 -mfpu=neon-fp-armv8 -ftree-vectorize -mvectorize-with-neon-quad -ftree-loop-vectorize -O2 -flto=8 -Wl,-flto=8 -fno-fat-lto-objects -ffast-math -funsafe-math-optimizations -ftrapv -Wno-error=format-nonliteral -Isrc/core/core@sta -Isrc/core -I../../../src/systemd/src/core -I. -I../../../src/systemd/ -Isrc/libsystemd-network -I../../../src/systemd/src/libsystemd-network -I../../../src/systemd/src/libsystemd/sd-network -I../../../src/systemd/src/libsystemd/sd-netlink -I../../../src/systemd/src/libsystemd/sd-id128 -I../../../src/systemd/src/libsystemd/sd-hwdb -I../../../src/systemd/src/libsystemd/sd-device -I../../../src/systemd/src/libsystemd/sd-bus -Isrc/libudev -I../../../src/systemd/src/libudev -Isrc/udev -I../../../src/systemd/src/udev -Isrc/login -I../../../src/systemd/src/login -Isrc/timesync -I../../../src/systemd/src/timesync -Isrc/resolve -I../../../src/systemd/src/resolve -Isrc/journal -I../../../src/systemd/src/journal -Isrc/systemd -I../../../src/systemd/src/systemd -Isrc/shared -I../../../src/systemd/src/shared -Isrc/basic -I../../../src/systemd/src/basic -I/home/asavah/kross/build/rpi3/rootfs/usr/include/libmount -I/home/asavah/kross/build/rpi3/rootfs/usr/include/blkid -I/home/asavah/kross/build/rpi3/rootfs/usr/include/uuid -flto -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=gnu99 -Wextra -Werror=undef -Wlogical-op -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wdeclaration-after-statement -Wfloat-equal -Wsuggest-attribute=noreturn -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wimplicit-fallthrough=5 -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Wdate-time -Wnested-externs -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -fPIE -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-format-signedness -Wno-error=nonnull -Werror=shadow -ffunction-sections -fdata-sections -include config.h -fPIC -pthread -MD -MQ 'src/core/core@sta/service.c.o' -MF 'src/core/core@sta/service.c.o.d' -o 'src/core/core@sta/service.c.o' -c ../../../src/systemd/src/core/service.c
In file included from ../../../src/systemd/src/basic/util.h:47:0,
                 from ../../../src/systemd/src/basic/hashmap.h:30,
                 from ../../../src/systemd/src/shared/bus-util.h:31,
                 from ../../../src/systemd/src/core/service.c:31:
../../../src/systemd/src/basic/missing.h:528:17: error: expected identifier before numeric constant
 #define MS_MOVE 8192
```

### In case of bug report: Steps to reproduce the problem

Build systemd with latest util-linux master @ https://github.com/karelzak/util-linux/commit/edc1c90cb972fdca1f66be5a8e2b0706bd2a4949 and linux headers 4.14.x (MS_* defines not fixed)

The possible cause is described here https://github.com/systemd/systemd/pull/8452 , defines clash ...

when building systemd with this line https://github.com/karelzak/util-linux/blob/master/libmount/src/libmount.h.in#L31 commented out the build succeeds.

ping @karelzak

Maybe a missing define guard somewhere ...

**Comment by karelzak** [comment]
It seems systemd src/basic/missing.h have the same problem we had in libmount.h :-)

If you have local fallbacks for MS_* constants than you have to include your stuff after sys/mount.h because sys/mount.h do not use #ifdefs for the constants.

The problem is visible because libmount.h from git tree includes sys/mount.h now.

I guess add "#include <sys/mount.h>" to the src/basic/missing.h will fix the problem, or you can remove the MS_* fallbacks from src/basic/missing.h at all as you duplicate libmount.h there.

**Comment by karelzak** [comment]
And note that sys/mount.h is so crazy that it uses 
```
enum
{  
  MS_RDONLY = 1,                /* Mount read-only.  */
#define MS_RDONLY       MS_RDONLY
```
to make it more complicated...

**Comment by asavah** [comment]
> I guess add "#include <sys/mount.h>" to the src/basic/missing.h will fix the problem, or you can remove the MS_* fallbacks from src/basic/missing.h at all as you duplicate libmount.h there.

Nope, tried that, build fails in different place now,
```
FAILED: src/basic/basic@sta/btrfs-util.c.o
armv8-rpi3-linux-gnueabihf-gcc -pipe -Wno-unused-value -Wno-unused-but-set-variable -Wno-deprecated-declarations -Wno-misleading-indentation -Wno-unused-function -Wno-unused-result -Wno-unused-parameter -Wimplicit-fallthrough=0 -Wno-unused-but-set-parameter -Wno-int-in-bool-context -mcpu=cortex-a53 -mfpu=neon-fp-armv8 -ftree-vectorize -mvectorize-with-neon-quad -ftree-loop-vectorize -O2 -flto=8 -Wl,-flto=8 -fno-fat-lto-objects -ffast-math -funsafe-math-optimizations -ftrapv -Wno-error=format-nonliteral -Isrc/basic/basic@sta -Isrc/basic -I../../../src/systemd/src/basic -I. -I../../../src/systemd/ -Isrc/libsystemd-network -I../../../src/systemd/src/libsystemd-network -I../../../src/systemd/src/libsystemd/sd-network -I../../../src/systemd/src/libsystemd/sd-netlink -I../../../src/systemd/src/libsystemd/sd-id128 -I../../../src/systemd/src/libsystemd/sd-hwdb -I../../../src/systemd/src/libsystemd/sd-device -I../../../src/systemd/src/libsystemd/sd-bus -Isrc/core -I../../../src/systemd/src/core -Isrc/libudev -I../../../src/systemd/src/libudev -Isrc/udev -I../../../src/systemd/src/udev -Isrc/login -I../../../src/systemd/src/login -I../../../src/systemd/src/time-wait-sync -Isrc/timesync -I../../../src/systemd/src/timesync -Isrc/resolve -I../../../src/systemd/src/resolve -Isrc/journal -I../../../src/systemd/src/journal -Isrc/systemd -I../../../src/systemd/src/systemd -Isrc/shared -I../../../src/systemd/src/shared -I/home/asavah/kross/build/rpi3/rootfs/usr/include/blkid -I/home/asavah/kross/build/rpi3/rootfs/usr/include/uuid -I/home/asavah/kross/build/rpi3/rootfs/usr/include/libmount -flto -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=gnu99 -Wextra -Werror=undef -Wlogical-op -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wdeclaration-after-statement -Wfloat-equal -Wsuggest-attribute=noreturn -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wimplicit-fallthrough=5 -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Wdate-time -Wnested-externs -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -fPIE -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-format-signedness -Wno-error=nonnull -Werror=shadow -ffunction-sections -fdata-sections -include config.h -fPIC -pthread -fvisibility=default -MD -MQ 'src/basic/basic@sta/btrfs-util.c.o' -MF 'src/basic/basic@sta/btrfs-util.c.o.d' -o 'src/basic/basic@sta/btrfs-util.c.o' -c ../../../src/systemd/src/basic/btrfs-util.c
In file included from ../../../src/systemd/src/basic/btrfs-util.c:24:0:
/home/asavah/kross/build/rpi3/rootfs/usr/include/sys/mount.h:35:3: error: expected identifier before numeric constant
   MS_RDONLY = 1,  /* Mount read-only.  */
   ^
```

**Comment by karelzak** [comment]
hmm... 

$ git grep "linux/fs.h"

src/basic/btrfs-util.c:#include <linux/fs.h>
src/basic/chattr-util.c:#include <linux/fs.h>
src/core/load-fragment.c:#include <linux/fs.h>
src/core/namespace.c:#include <linux/fs.h>
src/import/import-raw.c:#include <linux/fs.h>
src/import/import-tar.c:#include <linux/fs.h>
src/import/pull-raw.c:#include <linux/fs.h>
src/journal/journal-file.c:#include <linux/fs.h>
src/journal/journalctl.c:#include <linux/fs.h>
src/shared/machine-image.c:#include <linux/fs.h>
src/tmpfiles/tmpfiles.c:#include <linux/fs.h>

**Comment by karelzak** [comment]
@asavah  try this patch: https://github.com/karelzak/systemd/commit/eb50b6d936c474f5bc2974c059f6432222af4de4 It clean ups the includes in systemd. It works for me with latest util-linux as well as with old version.

**Comment by karelzak** [comment]
I did a small change to the libmount to make it more robust, anyway karelzak/systemd@eb50b6d is still a good idea IMHO.

**Comment by keszybz** [comment]
Should be fixed now with #8580.
